kep/4986-sakkara-cluster-topology/README.md

```diff
@@ -0,0 +1,536 @@
+  parallelism: 8
+  template:
+    metadata:
+      labels:
+        sakkara.group.name: "group-a"
+    spec:
+      schedulerName: sakkara
+      containers:
+      - name: main
+        image: busybox
+        imagePullPolicy: IfNotPresent
+        command: ["sh", "-c", "sleep 10"]
+        resources:
+          requests:
+            cpu: "100m"
+            memory: "128Mi"
+            nvidia.com/gpu: "1"
+          limits:
+            cpu: "200m"
+            memory: "128Mi"
+            nvidia.com/gpu: "1"
+      restartPolicy: OnFailure
+```
+
+In this case the pods of the group are specified in the pod template. For Sakkara, the following additions are needed.
+
+- *group name*
+  
+```yaml
+      labels:
+        sakkara.group.name: "group-a"
+```
+
+- *scheduler name*
+
+```yaml
+    spec:
+      schedulerName: sakkara
+```
+
+The group may be made up of a collection of pods, or a deployment, or multiple deployments, e.g. master and worker pods. The main requirement is that pods are homogeneous, i.e. all pods in the group have the same amount of requested resources.
+
+A group of pods without a corresponding group ConfigMap would remain in the pending state until the group ConfigMap is created.
+
+### Placement results
+
+In addition to the pods of a group being scheduled by Sakkara as a gang and according to topological constraints, some information about the placement result is provided in the group ConfigMap and the pods. This information may be used to configure the application containing the group.
+
+- *group placement*
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: group-a
+  namespace: default
+  labels: {
+    "sakkara.group": 
+  }
+data:
+  "name": "group-a"
+  "size": "8"
+  "priority": "0"
+  "constraints": |
+    {
+      "rack": {
+        "type": "pack"
+      },
+      "node": {
+        "type": "spread"
+      }
+    }
+  "status": Bound
+  "placement": |
+    {"root":
+      {"rack-0": 
+        {"node-0": {"pod-0", "pod-4", "pod-6"}},
+        {"node-2": {"pod-1", "pod-3", "pod-7"}}
+      },
+      {"rack-1": 
+        {"node-3": {"pod-2"}},
+        {"node-3": {"pod-5"}}
+      }
+    }
+  "rank": '[("pod-0",0) ("pod-4",1) ("pod-6",2) ("pod-1",3) ("pod-3",4) ("pod-7",5) ("pod-2",6) ("pod-5",7) ]'
+```
+
+Pods in the group are ranked based on their distance from a designated (master) pod. If a pod has the string "master" in its name, it will be considered the master pod, otherwise an arbitrary pod will be selected. Distance is the number of edges between two pods in the topology tree.
+
+- *pod placement*
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-6
+  labels:
+    sakkara.group.name: group-a
+    sakkara.member.rank: "2"
+    sakkara.member.retries: "1"
+    sakkara.member.status: Bound
+```
+
+The rank, status, and number of placement trials for the pod are provided as labels. This way, an (init) container in the pod may read the rank label and configure itself accordingly.
```

Comment:
It's an anti-pattern to update a pod's labels to store extra info. Update a pod is expensive, no need to say updating a gang of pods.

Reply:
The most important data is the rank. Having it as a pod label simplifies the logic of the container getting it initially as an environment variable. Alternatively, rank values for all pods in the group may be written in the CR for the group. The challenge is for the containers to get them?